beacon_chain/validators/beacon_validators.nim

```
@@ -166,8 +145,42 @@ proc addValidators*(node: BeaconNode) =
                                                  gasLimit)
     v.updateValidator(data)
 
-proc pollForDynamicValidators*(node: BeaconNode) {.async.} =
-  if node.config.validatorsSourceInverval == 0:
+proc addValidators*(node: BeaconNode) =
+  info "Loading validators", validatorsDir = node.config.validatorsDir(),
+                keystore_cache_available = not(isNil(node.keystoreCache))
+  let epoch = node.currentSlot().epoch
+
+  for keystore in listLoadableKeystores(node.config, node.keystoreCache):
+    let
+      data = withState(node.dag.headState):
+        getValidator(forkyState.data.validators.asSeq(), keystore.pubkey)
+      index =
+        if data.isSome():
+          Opt.some(data.get().index)
+        else:
+          Opt.none(ValidatorIndex)
+      feeRecipient = node.consensusManager[].getFeeRecipient(
+        keystore.pubkey, index, epoch)
+      gasLimit = node.consensusManager[].getGasLimit(keystore.pubkey)
+
+      v = node.attachedValidators[].addValidator(keystore, feeRecipient,
+                                                 gasLimit)
+    v.updateValidator(data)
+
+  try:
+    waitFor allFutures(mapIt(node.config.web3signers,
```

allFutures does not raise on error, you need to investigate each future individually - any exceptions raised will be ignored / discarded

---

also, the `waitFor` is potentially problematic if this function runs in a context of another loop - not sure we have a good way of preventing that - it would be better to make the function async and eliminate this risk

---

This code is a simple refactoring of the previous version which was also using `waitFor` here (querying a single source). The use of `allFutures` was intentional to some extent because `addValidatorsFromWeb3Signer` handles all possible exceptions (having additional sanity checks here doesn't hurt of course).

---

well, the sanity checks in this case also swallow cancellation (and imply that error handling is written in a schematic rather than deliberate way) - I'm not sure that was intended - and because allFutures doesn't cancel its dependent futures, a cancellation here looks like it will leak

---

ah, no cancellation either, my bad - perhaps we should log an error though since an exception leaking out of that construct this would imply a bug somewhere